### ISSUE TITLE
perf(cuda): f16 activation policy below Ampere (bf16 has no Volta hardware path)

### ISSUE THREAD

**Issue by inureyes** [lablup/mlxcel#1542]
Part of #1536. Phase 2. Depends on #1537 for the capability probe and #1538 for the baseline. **Gates #1543** — Volta tensor cores are f16-only, so there is no tensor-core path for a bf16 activation graph at all.

This is the policy decision of the program. It is deliberately its own issue because it is the one item with a real quality risk, and because it reverses a patch that is correct on the architectures mlxcel actually ships for.

## Context

`src/lib/mlx-cpp/patches-cuda/dtype.cpp` is a CUDA-only overlay that changes MLX's type-promotion table so bf16 survives contact with float32:

> bf16 + float32 -> bf16 (was float32) ... On CUDA, bf16 is the native compute type for LLM inference. The standard JAX promotion rules trigger `copy_v<bf16, float>` conversion kernels at every binary operation boundary, causing ~58% GPU time overhead on 9B models.

That reasoning is sound on GB10 and every other arch in the release matrix. It is precisely inverted on Volta, where bf16 is not a compute type at all:

- **No bf16 ALU.** sm_70 predates `__nv_bfloat16` arithmetic, which requires `__CUDA_ARCH__ >= 800`. Every bf16 op is emulated through float conversion.
- **No bf16 tensor core.** `cute/arch/mma_sm70.hpp` provides only `SM70_8x8x4_F16F16F16F16_*` and `SM70_8x8x4_F32F16F16F32_*`. There is no bf16 Volta MMA and there cannot be one.
- **No cuBLAS bf16 GEMM.** `CUDA_R_16BF` support begins at sm_80.

f16, by contrast, is fully supported on Volta in all three: native `__hfma2`, `SM70_8x8x4` tensor-core MMA, and cuBLAS `HGEMM`. Meanwhile the checkpoints in play are bf16 — `qwen3.8-27B-4bit` carries 709 BF16 tensors alongside its 138 U32 quantized planes — and the #1538 profile is full of `__nv_bfloat16` elementwise, norm, binary, and copy kernels, every one of them paying emulation.

## Scope

Make the activation dtype policy arch-aware: keep today's bf16 preservation on sm_80+, and route pre-Ampere to f16.

Out of scope: the checkpoint format on disk (no re-quantization, no new artifacts), and Metal, which uses the unmodified upstream `dtype.cpp` and must stay untouched.

## Design decision to make first

Two viable routes; pick one with evidence, do not implement both.

**(a) Load-time cast.** Convert bf16 tensors to f16 as the loader materializes them on a pre-Ampere device. Contained, explicit, and visible in `MLXCEL_TRACE_DTYPE` output. Costs a one-time conversion at load and leaves the promotion table alone.

**(b) Arch-conditional promotion table.** Compile the `patches-cuda/dtype.cpp` rules so pre-Ampere promotes toward f16 instead of preserving bf16. Cheaper at load, but `dtype.cpp` is a single translation unit shared by one binary that may target several archs in one fat build, so the rule would have to be a runtime branch on the capability probe rather than a compile-time one — and a promotion table that varies at runtime is a large blast radius for a rule the rest of MLX assumes is static.

(a) is the recommended default on blast-radius grounds. State the choice and the reason in the issue before writing code.

## Numerical risk, and how it is bounded

f16 has a 5-bit exponent (max 65504) against bf16's 8-bit fp32-range exponent. This is the real risk of the issue and the reason it is separated from Phase 1:

- Pre-softmax attention logits, residual-stream accumulation in deep stacks, and some MLP intermediates are the known overflow sites.
- Mitigations already present in the codebase: RMSNorm and softmax accumulate in f32 (`rms_norm_small` reduces through `float`), and the KV cache is already f16 (`KV cache mode: fp16 (applied to 64 of 64 layers)` on the baseline run), so KV storage introduces no new exposure.
- This model additionally sets `mamba_ssm_dtype = float32` for the GDN recurrence, which stays f32 either way.

The acceptance criteria below require overflow to be *measured*, not assumed absent.

## Implementation plan

1. Decide (a) vs (b) and record the rationale.
2. Gate on the #1537 capability probe: pre-Ampere only. sm_80+ must be untouched, and the `patches-cuda/dtype.cpp` behavior there must be bit-for-bit what it is today.
3. Add an escape hatch following the conventions in `docs/environment-variables.md`: `MLXCEL_CUDA_ACTIVATION_DTYPE` accepting `auto` (default), `f16`, `bf16`, so a Volta user can A/B and roll back without a rebuild, and so the overflow investigation has a control arm.
4. Instrument for overflow: a debug-gated non-finite check on layer outputs during a validation sweep, reporting the first layer and tensor to produce a non-finite value. This does not ship enabled.
5. Re-run the `MLXCEL_TRACE_ASTYPE` conversion-node count (issue #636's metric) on Volta before and after. The original patch existed to eliminate `AsType` nodes; confirm the f16 route does not reintroduce a conversion storm of its own.

## Acceptance criteria

- [ ] Route (a) or (b) chosen with a written rationale.
- [ ] Decode and prefill tok/s on Volta before and after, against #1538, with the nsys kernel table showing `__nv_bfloat16` kernels replaced by `__half` ones.
- [ ] `MLXCEL_TRACE_ASTYPE` node count on Volta does not regress against the bf16 arm.
- [ ] **Quality, not just speed**: perplexity on the `wikitext2` excerpt already in `tests/fixtures/wikitext2_excerpt.txt` for f16 vs bf16 on Volta, at 128/512/2048-token windows. A regression beyond noise blocks the issue.
- [ ] Zero non-finite values across the perplexity sweep and a 2048-token generation on at least two model families (one dense-hybrid, one MoE). Any overflow site found is documented with the layer name, whether or not it is mitigated.
- [ ] `MLXCEL_CUDA_ACTIVATION_DTYPE=bf16` restores the current behavior exactly on Volta.
- [ ] **Zero change on sm_80+**: GB10 greedy output byte-identical, GB10 baseline throughput unmoved.
- [ ] Metal unaffected — `patches-cuda/` is CUDA-only by construction; assert it in review.
- [ ] Changelog entry states that Volta activations are f16 and that Volta output is therefore not comparable token-for-token with Ampere+.

## Validation

```
MLX_CUDA_ARCHITECTURES=70 make release-cuda
for d in auto bf16; do MLXCEL_CUDA_ACTIVATION_DTYPE=$d ./target/release/mlxcel generate \
  -m ./models/qwen3.8-27B-4bit -p "Hi." -n 40; done
MLXCEL_TRACE_ASTYPE=2 ./target/release/mlxcel generate -m ./models/qwen3.8-27B-4bit -p "Hi." -n 4
```

## References

- The patch being made arch-conditional: `src/lib/mlx-cpp/patches-cuda/dtype.cpp` (rationale in its header comment).
- Overlay gating that keeps it CUDA-only: `src/lib/mlx-cpp/CMakeLists.txt:66-80`.
- Volta MMA atoms, showing the absence of any bf16 variant: `cute/arch/mma_sm70.hpp` in the vendored CUTLASS.
- `AsType` node-count metric and its variable: #636, `docs/environment-variables.md` (`MLXCEL_TRACE_ASTYPE`).
